simulator in hardware, new module

romain-chiap opened the pull request:
new module added that allows to run a simulator for quadrotor directly on the hardware autopilot.
For more documentation visit 
https://github.com/romain-chiap/PX4_SIH_QuadX

Please use [PX4 Discuss](http://discuss.px4.io/) or [Slack](http://slack.px4.io/) to align on pull requests if necessary.

**Test data / coverage**
https://review.px4.io/plot_app?log=4a5b5b89-7231-4e7e-a989-f16e71aa9966

**Describe problem solved by the proposed pull request**
Timing delays due to the bidirectional connection from the hardware to the desktop computer are an issue in the HITL and affect the flight performance. Furthermore, a user willing to modify the quadrotor parameters (mass, inertia, maximum thrust, ...) needs to dig into JMAVSim or Gazebo to do the modification.

**Describe your preferred solution**
The proposed solution is to add a new module, fully compatible with PX4, called Simulator In Hardware (SIH). It implements the equations of motion of a quadrotor along with thruster force and torque, a simple drag model is also implemented. The sih module subscribes to actuator_outputs to get directly the outputs of the mixer. It also reconstructs noisy sensor data and publish them. The idea is to include the state estimator in the loop, and the controller of course.

The user can modify the drag coefficient, inertia matrix and mass of the vehicle directly from QGC parameters. 

This module is fully compatible with the current PX4 Firmware. The following flight modes have been tested:
- Manual
- Altitude
- Accro
- Mission
- Failsafe (RC signal lost)

**Describe possible alternatives**
The other alternatives are to use the SITL or current HITL

**Additional context**
This module was developed to display the quadrotor trajectory in FlightGear, to have a real time visual.
This module outputs the quad position and orientation on a serial port. The idea is to transform those serial data into UDP to communicate the position/attitude to FlightGear for the display. A Java application SerialToUDP.jar is also provided at the following link along with a simple quadrotor model for FlightGear
https://github.com/romain-chiap/PX4_SIH_QuadX

dagar commented:
Cool! Let me dig into this some more and we can discuss it.

A few initial minor things that stand out are including it on more boards (should be almost everywhere except fmu-v2, io-v2, etc), a mechanism to start it (via param) with the right airframe (safe configuration), and metadata for the parameters.

bresch commented:
It's a really interesting approach. Traditionally, HIL is used because the flight simulator is way too heavy to run on the autopilot itself.

I'll try and give you some feedback as soon as possible!

LorenzMeier commented:
Super cool!

romain-chiap commented:
I formatted the code with astyle and the metadata for the parameters.

Regarding the board configs, mc_att_control appears in 40 different cmake files, so I guess I should include it in the same files except:
1. /px4/fmu-v2/*.cmake
1. /px4/sitl/*.cmake 
1. any other ?

I think I should exclude it from SITL (to avoid an Inception like simulator :). Anyway, the sensors data will be conflicting with the data from SITL, so I won't include it.

I will check to add another airframe but I'm not very familiar with that.

MaEtUgR commented on src/modules/sih/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+############################################################################
+#
+#   Copyright (c) 2015 PX4 Development Team. All rights reserved.
```
nit-pick: you could update all the years in the headers

MaEtUgR commented:
Devcall: Cool feature! Can we have a minimal automatic test to run in CI and make sure it doesn't break in the future? Can we add some documentation?

@dagar : Suggestion to use jMAVsim or gazebo to display the vehicle state. By using this [MAVLink message](https://mavlink.io/en/messages/common.html#HIL_STATE_QUATERNION) which get sent [here](https://github.com/PX4/Firmware/blob/12d29fd093e3e988308d0b7b027d7a45eeaf5b5c/src/modules/mavlink/mavlink_messages.cpp#L4597-L4688) for the ground truth and having a mode for the simulator to just display the state instead of simulating.

romain-chiap commented:
done:
1. sih module added in most board configs
1. new airframe added 
    - I added a parameter SYS_SIH, which is set to 1 by the airframe
    - rcS starts the sih module if SYS_HITL and SYS_SIH equal 1
1. HIL_STATE_QUATERNION sent through MAVLink
    - I made a correction in [mavlink_messages.cpp](https://github.com/PX4/Firmware/pull/11400/commits/0d1490206541dd391b5cb8896146a865e4d460af#diff-e605251d3710c0f7f0c065b242004e9c) to match the units described in common.xml

in progress:
1. read HIL_STATE_QUATERNION on jMAVSim
1. real-time callback for sih 
    - I tried with hrt_call_every() to call the [inner loop](https://github.com/PX4/Firmware/pull/11400/commits/0d1490206541dd391b5cb8896146a865e4d460af#diff-cfeeba33c0520fafd8c957221802849bR228) (which itself publishes ORB topics) but I obtained `orb_publish failed`. Is it a problem to publish ORB topics from a callback? Is it because the priority is set to maximum?
    - the inner loop takes about 420 us to be executed

romain-chiap commented:
> Devcall: Cool feature! Can we have a minimal automatic test to run in CI and make sure it doesn't break in the future? Can we add some documentation?

Where can I add the documentation? This is currently in a [pdf file](https://github.com/romain-chiap/PX4_SIH_QuadX/blob/master/Documentation/Dynamic%20Model%20and%20Parameters.pdf)

MaEtUgR commented:
@romain-chiap Probably docs should go to the development guide's simulation part:
https://dev.px4.io/en/simulation/
https://github.com/PX4/Devguide/tree/master/en/simulation
You could create a page for your onboard simulation. Maybe @hamishwillee should confirm first.

hamishwillee commented on src/lib/systemlib/system_params.c:
```diff
@@ -77,6 +77,19 @@ PARAM_DEFINE_INT32(SYS_AUTOCONFIG, 0);
+ * Enable SIH mode on next boot
+ *
+ * The simulation in hardware (SIH) will enable a quad simulator to run on the autopilot.
+ * When disabled the same vehicle can be normally flown outdoors.
```
Only outdoors?

Perhaps:
```
 * Enable Simulation in Hardware (SIH) mode on next boot.
 *
 * SIH mode enables Hardware in the Loop (HIL) quadcopter simulation where the simulator runs on the flight controller itself.
 * When disabled the same vehicle can be normally.
```

hamishwillee commented:
@bkueng 
What category and subcategory should this be in for the module documentation? Currently this fails because a custom category was used:
```
 make module_documentation
ninja: Entering directory `/home/ubuntu/src/Firmware/build/px4_sitl_default'
[0/1] Generating module documentation
Scanning source path ['/home/ubuntu/src/Firmware/src']
Exception in file /home/ubuntu/src/Firmware/src/modules/sih/sih.cpp
Traceback (most recent call last):
  File "/home/ubuntu/src/Firmware/Tools/px_process_module_doc.py", line 104, in <module>
    main()
  File "/home/ubuntu/src/Firmware/Tools/px_process_module_doc.py", line 84, in main
    if not scanner.ScanDir(args.src_path, parser):
  File "/home/ubuntu/src/Firmware/Tools/px4moduledoc/srcscanner.py", line 24, in ScanDir
    if not self.ScanFile(path, parser):
  File "/home/ubuntu/src/Firmware/Tools/px4moduledoc/srcscanner.py", line 49, in ScanFile
    return parser.Parse(scope, contents)
  File "/home/ubuntu/src/Firmware/Tools/px4moduledoc/srcparser.py", line 304, in Parse
    module_doc.category() + ' for ' + scope)
Exception: Invalid/unknown category sih for modules/sih
FAILED: CMakeFiles/metadata_module_documentation 
```

hamishwillee commented on src/lib/systemlib/system_params.c:
```diff
@@ -77,6 +77,19 @@ PARAM_DEFINE_INT32(SYS_AUTOCONFIG, 0);
 PARAM_DEFINE_INT32(SYS_HITL, 0);
 
+/**
+ * Enable SIH mode on next boot
```
@bkueng Should this be with all the other SIH parameters? I do appreciate this is enabling new system behaviour, but I would have thought it better to group with the associated module?

bkueng replied:
I guess I prefer extending SYS_HITL=2 instead of adding another parameter for this. Given that we need to disable sensors in both cases.

romain-chiap replied:
I am not against setting SYS_HITL=2. 

[rcS](https://github.com/PX4/Firmware/pull/11400/files#diff-e5c16c5209413f30e964883f74027b51L327) will need to be modified as
```suggestion
	if param greater SYS_HITL 0
	then
		set OUTPUT_MODE hil
		sensors start -h
		commander start --hil
		# disable GPS
		param set GPS_1_CONFIG 0

		# start the simulator in hardware if needed
		if  param compare SYS_HITL 2
		then
			sih start
		fi
```
and [rc.interface](https://github.com/PX4/Firmware/blob/master/ROMFS/px4fmu_common/init.d/rc.interface#L205) will need to be modified as
```suggestion
if param greater SYS_HITL 0
```
of course, the comments on [system_params.c](https://github.com/PX4/Firmware/pull/11400/files/8206d6c73abe28ae4720d25902efbd2cd66ec98e#diff-58f0a6ba05be221eaead7a8dd240c391L68) will need to be updated 
```suggestion
/**
 * Enable HITL mode on next boot
 *
 * While enabled the system will boot in HITL mode and not enable all sensors and checks.
 * When disabled the same vehicle can be flown normally.
 * Setting SYS_HITL to 1 will enable the hardware in the loop.
 * Setting SYS_HITL to 2 will enable the simulation in hardware (SIH).
 *
 * @boolean
 * @reboot_required true
 *
 * @group System
 */
PARAM_DEFINE_INT32(SYS_HITL, 0);
```

Any other file using SYS_HITL, that will need to be modified?

bkueng replied:
> Any other file using SYS_HITL, that will need to be modified?

That looks correct, except for the param defintion. It should be something like:
```
/**
 * Enable HITL/SIH mode on next boot
 *
 * While enabled the system will boot in Hardware-In-The-Loop (HITL) or Simulation-In-Hardware (SIH) mode and not enable all sensors and checks.
 * When disabled the same vehicle can be flown normally.
 *
 * @value 0 HITL disabled
 * @value 1 HITL enabled
 * @value 2 SIH enabled
 * @reboot_required true
 *
```

hamishwillee commented on src/modules/sih/sih_params.c:
```diff
@@ -0,0 +1,345 @@
+ * @min 0.0
+ * @decimal 2
+ * @increment 0.1
+ * @group SIH
```
Perhaps the group name should be **Simulation In Hardware**. At this point the acronym isn't as well known as SITL or HITL so while eventually SIH might be obvious, right now it is not.

romain-chiap replied:
I put two underscores to:
- SIH__LAT0
- SIH__LON0
- SIH__H0
- SIH__MU_X
- SIH__MU_Y
- SIH__MU_Z

I would like them to be grouped in the documentation. Is there a better way to do it?

romain-chiap replied:
I'm thinking of removing those six parameters as they are very specific and don't bring much if changed by the user.

hamishwillee commented on ROMFS/px4fmu_common/init.d/airframes/1100_rc_quad_x_sih.hil:
```diff
@@ -0,0 +1,21 @@
+#
+# @name SIH Quadcopter X
+#
+# @type Simulation
```
FYI, you can see how this will render here: https://hamishwillee.gitbooks.io/havdevguide/content/v/test_sih/en/airframes/airframe_reference.html#simulation-copter

romain-chiap replied:
great!

hamishwillee commented:
@romain-chiap Very very cool. A few comments inline.

With respect to documentation, as @MaEtUgR says above, it should appear in the table here: http://dev.px4.io/en/simulation/#supported-simulators and it should have its own sub page linking to it. 

The sub page should provide overview information in the same way as the other pages:
- What is the simulator
- A diagram showing the architecture
- Why is it "special"/good
- How to set it up and run it 

Mostly this is what you have in https://github.com/romain-chiap/PX4_SIH_QuadX/wiki/Setting-up-the-Coriolis-SIH and on your readme.

You would also link to the generated parameters: https://hamishwillee.gitbooks.io/havdevguide/content/v/test_sih/en/advanced/parameter_reference.html#sih
And to your project page if you wish (I say if you wish because since this is now in PX4 codeline best not to jump people to too many places).

The PDF you link above is pretty cool (though we generate the params, so that section not so useful). It is up to you if you include that content in our docs. IMO that is quite a lot of work so ideally you would copy the file into https://github.com/PX4/Devguide/tree/master/assets/simulation and link to it from there.

I don't know how well you know gitbook, but that is what we use for our infrastructure. Essentially you create files in markdown and submit a PR to https://github.com/PX4/Devguide. Info on the toolchain is here: http://dev.px4.io/en/contribute/docs.html#adding-new-content---big-changes
Don't stress too much about this - your existing docs look pretty good and I can subedit as needed for you.

Is that enough to get started?

romain-chiap commented:
@hamishwillee 
Thanks, I appreciate the comments.
I'm still playing with jMAVSim to have a visualization for the quad. This will need to be described in the documentation as well.
In the meantime I will have a look at gitbook and eventually open a PR when ready.

hamishwillee commented:
Great. Minimally you could draft a link to your docs and then update when you have a little more time.

bkueng commented on msg/CMakeLists.txt:
```diff
@@ -107,6 +107,7 @@ set(msg_files
 	sensor_preflight.msg
 	sensor_selection.msg
 	servorail_status.msg
+	sih.msg
```
I'd prefer a more descriptive name, like `hardware_simulator_data`?

romain-chiap replied:
The sih messages are just for debug purpose (i.e. for me).
The information contains in sih.msg is a combination of the 3 messages:
- actuator_outputs
- vehicle_global_position_groundtruth
- vehicle_attitude_groundtruth

So if needed I can remove sih.msg.

mhkabir replied:
Let's remove it before merge in that case.

romain-chiap replied:
done

bkueng commented on src/modules/mavlink/mavlink_messages.cpp:
```diff
@@ -4662,12 +4662,13 @@ class MavlinkStreamGroundTruth : public MavlinkStream
-			msg.vy = gpos.vel_e;
-			msg.vz = gpos.vel_d;
+			// same units as defined in mavlink/common.xml
+			msg.lat = gpos.lat * 1e7;
```
This was indeed incorrect. Thanks for the fix.

hamishwillee commented on src/modules/sih/sih.cpp:
Second param should be "simulation" as per https://github.com/PX4/Firmware/pull/11400#pullrequestreview-214969363 (which also shows where you'll need to update the category in module docs tools)

MaEtUgR commented on src/modules/sih/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+		sih.cpp
+	DEPENDS
+		mathlib
+	)
```
nit-pick: newline

MaEtUgR replied:
nit-pick: no newline at the end of the file

MaEtUgR commented on src/modules/sih/sih.cpp:
newline

MaEtUgR commented on src/modules/sih/sih.cpp:
newline

MaEtUgR commented on src/modules/sih/sih_params.c:
```diff
@@ -0,0 +1,345 @@
+ * @increment 0.001
+ * @group Simulation In Hardware
+ */
+PARAM_DEFINE_FLOAT(SIH__MU_Z,  0.504f);
```
newline

romain-chiap commented:
@hamishwillee 
I created [PR757](https://github.com/PX4/Devguide/pull/757) on the Devguide.

bkueng commented on src/modules/sih/sih.cpp:
Can you remove these?

bkueng commented on src/modules/sih/sih.cpp:
Can you cancel before destroying the semaphore?

romain-chiap commented on src/modules/sih/sih.cpp:
@bkueng 
the 5 lines (85-89)?
Shouldn't I keep `PRINT_MODULE_USAGE_NAME("sih", "simulation");`?

bkueng replied:
Not all, just the example flags that this module doesn't use.

romain-chiap commented on src/modules/mavlink/mavlink_main.cpp:
```diff
@@ -824,12 +824,16 @@ Mavlink::set_hil_enabled(bool hil_enabled)
 		_hil_enabled = true;
 		ret = configure_stream("HIL_ACTUATOR_CONTROLS", 200.0f);
+
+		configure_stream("GROUND_TRUTH", 200.0f); 	// HIL_STATE_QUATERNION to display the SIH
```
I was displaying it on jMAVSim and the rendering was laggy, so I was playing with that value. But 25 Hz will work fine

romain-chiap replied:
BTW is there a way to check that SYS_HITL == 2 ?
to avoid sending GROUND_TRUTH during traditional HITL simulation.

bkueng replied:
We would need to read the parameter, which we can do. But currently mavlink gets the HIL information via vehicle_status.hil_state message.

romain-chiap commented on src/modules/sih/sih_params.c:
```diff
@@ -0,0 +1,345 @@
+ * @max  850000000
+ * @group Simulation In Hardware
+ */
+PARAM_DEFINE_INT32(SIH__LAT0, 454671160);
```
I would like those six parameters to appear grouped in the documentation

bkueng replied:
Ok, your trick works, but I'd use a common prefix for that, like `SIH_LOC_xy`.

romain-chiap commented on ROMFS/px4fmu_common/init.d/rcS:
```diff
@@ -324,15 +324,20 @@ else
 		param set GPS_1_CONFIG 0
 
+		# start the simulator in hardware if needed
+		if  param compare SYS_HITL 2
```
@bkueng 
which file shall I run? 
I ran `make format`, but it does not seem to format rcS.

bkueng replied:
`make format` is correct, but it won't check the rcS files.

romain-chiap commented:
> Devcall: Cool feature! Can we have a minimal automatic test to run in CI and make sure it doesn't break in the future?

@MaEtUgR 
Shall we discuss that in the next devcall (tomorrow)? I have no idea how to create an automatic test.

hamishwillee commented:
Note there are also file conflicts above to be fixed.

bkueng commented on src/modules/sih/sih.cpp:
One more left.

bkueng commented on src/modules/sih/sih.cpp:
```suggestion
Sih::Sih()
```

romain-chiap commented:
I rebased from PX4/Firmware master.
1. I [synchronized](https://digitaldrummerj.me/git-syncing-fork-with-original-repo/) my Fork with PX4/Firmware
1. I [rebased](https://git-scm.com/docs/git-rebase) my branch simulator_in_hardware from my branch master
1. I manually solved the conflicts (i.e. srcparser.py)
1. I updated the submodules
1. I compiled and checked the SIH was working properly

bkueng commented:
Something did not work as expected (you can see it from the many commits that now appear in the PR). Can you try to rebase again?

romain-chiap commented:
@bkueng 
I need some help with rebase (or undo rebase first).

What I did:
1. checking where I should reset `git reflog`
1. resetting to the chosen HEAD ` git reset --hard HEAD@{52}`
1. `make distclean`
1. commit
1. `git push --force`

It looks good on [my repository branch](https://github.com/romain-chiap/Firmware/commits/simulator_in_hardware), but for some reason, it did not followed to the PR.

Is there a good way to fix that? (or shall I make a new "clean" PR?)

bkueng commented:
It looks good now, but unfortunately, there are conflicts again.
Can you rebase again, and look into CI? For example I see this failure:
```
../../src/modules/sih/sih.hpp:149:14: fatal error: private field '_counter' is not used [-Wunused-private-field]

        int32_t     _counter = 0;
```
